common/src/main/java/com/linecorp/centraldogma/common/MergeQuery.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.centraldogma.common;
+
+import static com.linecorp.centraldogma.common.QueryType.IDENTITY;
+import static com.linecorp.centraldogma.common.QueryType.JSON_PATH;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * A merge query on files.
+ *
+ * @param <T> the content type of files being queried
+ */
+public interface MergeQuery<T> {
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     */
+    static MergeQuery<JsonNode> ofJson(MergeSource... mergeSources) {
+        return ofJson(ImmutableList.copyOf(requireNonNull(mergeSources, "mergeSources")));
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     */
+    static MergeQuery<JsonNode> ofJson(Iterable<MergeSource> mergeSources) {
+        return new JsonMergeQuery(IDENTITY, mergeSources, ImmutableList.of());
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}. Then, the specified
+     * <a href="https://github.com/json-path/JsonPath/blob/master/README.md">JSON path expressions</a>
+     * are applied to the content.
+     *
+     * @param mergeSources the paths of JSON files being merged and indicates whether it is optional
+     * @param jsonPaths the JSON path expressions to apply
+     */
+    static MergeQuery<JsonNode> ofJsonPath(Iterable<MergeSource> mergeSources,
+                                           String... jsonPaths) {
+        return ofJsonPath(mergeSources, ImmutableList.copyOf(requireNonNull(jsonPaths, "jsonPaths")));
+    }
+
+    static MergeQuery<JsonNode> ofJsonPath(Iterable<MergeSource> mergeSources,
+                                           Iterable<String> jsonPaths) {
+        return new JsonMergeQuery(JSON_PATH, mergeSources, jsonPaths);
+    }
+
+    /**
+     * Returns a newly-created {@link MergeQuery} that merges the JSON contents as specified in the
+     * {@code mergeSources}. Then, the specified
+     * <a href="https://github.com/json-path/JsonPath/blob/master/README.md">JSON path expressions</a>
+     * are applied to the content.
```

Comment:
Please generalize this just in case we add more `QueryType`s than just JSON path.